### ISSUE TITLE
loss is nan when pretaining on my own dataset

### ISSUE THREAD

**Issue by liangzimei** [uta-smile/TCL#4]
hi,  thanks for your excellent work firstly. when i train my own chinese dataset (so i change the bert-base-uncased to bert-base-chinese), loss becomes nan after several iterations. i have tried to decrease the lr and add grad_clip, but the problem  still  exists.
![image](https://user-images.githubusercontent.com/12395126/161052202-6deca95a-27de-46b1-b6db-0ad453d36108.png)
here is my training config:
![image](https://user-images.githubusercontent.com/12395126/161057031-5bab4833-61f9-4f3e-9b0a-2cc77fdb4358.png)

can you give me some suggestion? thanks in advance.

**Comment by viyjy** [comment]
Hi, thanks for your interest in our paper.
Could you please use the following command line for pre-training?
```
python -m torch.distributed.launch --nproc_per_node=8 \
--use_env Pretrain.py \
--config ./configs/Pretrain.yaml \
--output_dir output/pretrain \
--text_encoder bert-base-chinese
``` 
More details can be found in this [line](https://github.com/uta-smile/TCL/blob/b6d8a962e43fe7f645b7b4d93315fce619efd055/Pretrain.py#L190)
Please let me know if this helps. Thanks.

**Comment by viyjy** [comment]
Sorry, I just realized that you are using `bert-base-chinese`.
Can you show me `config_bert_chinese.json`? Thanks.

**Comment by liangzimei** [comment]
> Sorry, I just realized that you are using `bert-base-chinese`. Can you show me `config_bert_chinese.json`? Thanks.
 my config_bert_chinese.json is like this:
{
    "architectures": [
      "BertForMaskedLM"
    ],
    "attention_probs_dropout_prob": 0.1,
    "directionality": "bidi",
    "hidden_act": "gelu",
    "hidden_dropout_prob": 0.1,
    "hidden_size": 768,
    "initializer_range": 0.02,
    "intermediate_size": 3072,
    "layer_norm_eps": 1e-12,
    "max_position_embeddings": 512,
    "model_type": "bert",
    "num_attention_heads": 12,
    "num_hidden_layers": 12,
    "pad_token_id": 0,
    "pooler_fc_size": 768,
    "pooler_num_attention_heads": 12,
    "pooler_num_fc_layers": 3,
    "pooler_size_per_head": 128,
    "pooler_type": "first_token_transform",
    "type_vocab_size": 2,
    "vocab_size": 21128,
    "fusion_layer": 6,
    "encoder_width": 768
  }

**Comment by viyjy** [comment]
How many GPUs are you using?

**Comment by liangzimei** [comment]
> How many GPUs are you using?

i also use 8 gpus. 
and today  i use the same configs and dataset to run ALBEF, the loss looks normal now.
![image](https://user-images.githubusercontent.com/12395126/161187192-0d5fd428-bb9c-46f4-9093-2fa456e775b6.png)

**Comment by viyjy** [comment]
Thanks. May I know which dataset are you using? In this case, I need to test it on my machine to reproduce the problem.

**Comment by liangzimei** [comment]
> Thanks. May I know which dataset are you using? In this case, I need to test it on my machine to reproduce the problem.

sorry, the data is selected by myself, mainly from '抖音' app. the image-text pairs contain the videos' title and cover.

**Comment by viyjy** [comment]
Thanks. I will try to collect some data with chinese text to figure out the problem. Will let you know ASAP.

**Comment by liangzimei** [comment]
> > Thanks. May I know which dataset are you using? In this case, I need to test it on my machine to reproduce the problem.
> 
> sorry, the data is selected by myself, mainly from '抖音' app. the image-text pairs contain the videos' title and cover.

Or i can upload a portion of data to google drive ?  is it convenient for you?

**Comment by viyjy** [comment]
> > > Thanks. May I know which dataset are you using? In this case, I need to test it on my machine to reproduce the problem.
> > 
> > 
> > sorry, the data is selected by myself, mainly from '抖音' app. the image-text pairs contain the videos' title and cover.
> 
> Or i can upload a portion of data to google drive ? is it convenient for you?

Sure, I appreciate it. You can email the data link to jinyu.yang@mavs.uta.edu

**Comment by liangzimei** [comment]
> > > > Thanks. May I know which dataset are you using? In this case, I need to test it on my machine to reproduce the problem.
> > > 
> > > 
> > > sorry, the data is selected by myself, mainly from '抖音' app. the image-text pairs contain the videos' title and cover.
> > 
> > 
> > Or i can upload a portion of data to google drive ? is it convenient for you?
> 
> Sure, I appreciate it. You can email the data link to [jinyu.yang@mavs.uta.edu](mailto:jinyu.yang@mavs.uta.edu)
ok, i have alreadly sent the data to you ~

**Comment by viyjy** [comment]
Hi, the reason of this nan loss is that your dataset contains empty captions, for example one data sample in your dataset is `{'caption': '', 'image': '6912609964733271309_00001.jpg'}`. Removing such invalid samples solves this problem. 
You can use the following code to remove invalid samples from the json file.
```
json_path = 'data.json'
new_json_path = 'data_new.json'

f = open(json_path)
data = json.load(f)
new_data = []

for i in range(len(data)):
    if len(data[i]['caption'].strip()) != 0:
        new_data.append(data[i])

f.close()

with open(new_json_path, 'w') as jsonfile:
    json.dump(new_data, jsonfile, ensure_ascii=False)
```

Feel free to let me know if you might need additional information. Thanks.

**Comment by liangzimei** [comment]
> Hi, the reason of this nan loss is that your dataset contains empty captions, for example one data sample in your dataset is `{'caption': '', 'image': '6912609964733271309_00001.jpg'}`. Removing such invalid samples solves this problem. You can use the following code to remove invalid samples from the json file.
> 
> ```
> json_path = 'data.json'
> new_json_path = 'data_new.json'
> 
> f = open(json_path)
> data = json.load(f)
> new_data = []
> 
> for i in range(len(data)):
>     if len(data[i]['caption'].strip()) != 0:
>         new_data.append(data[i])
> 
> f.close()
> 
> with open(new_json_path, 'w') as jsonfile:
>     json.dump(new_data, jsonfile, ensure_ascii=False)
> ```
> 
> Feel free to let me know if you might need additional information. Thanks.


Thank you very much, it is my mistake.

**Comment by caisarl76** [comment]
> Hi, the reason of this nan loss is that your dataset contains empty captions, for example one data sample in your dataset is `{'caption': '', 'image': '6912609964733271309_00001.jpg'}`. Removing such invalid samples solves this problem. You can use the following code to remove invalid samples from the json file.
> 
> ```
> json_path = 'data.json'
> new_json_path = 'data_new.json'
> 
> f = open(json_path)
> data = json.load(f)
> new_data = []
> 
> for i in range(len(data)):
>     if len(data[i]['caption'].strip()) != 0:
>         new_data.append(data[i])
> 
> f.close()
> 
> with open(new_json_path, 'w') as jsonfile:
>     json.dump(new_data, jsonfile, ensure_ascii=False)
> ```
> 
> Feel free to let me know if you might need additional information. Thanks.

I was stuck at the same error and this helped me out! 
Thanks for @viyjy and @liangzimei too 😄
